website/docs/development/writing-tests.md

```diff
@@ -10,29 +10,237 @@ your controller in `test/apivalidations`.
+
+Very similar to the `create-minimal` test, except it sets all available fields
+of the resource spec.
+
+Sometimes you'll find options that are mutually exclusive. In that case you may
+need to write separate tests to exercise all the options. Try to come up with
+a meaningful name for the scenario, for example `create-sriov`.
+
+All controllers should implement this test.
+
+#### dependency
+
+Whenever your controller has dependencies on ORC resources, we want to ensure
+it is able to wait for the resource to exist and be ready, and optionally have
+a dependency guard on the resource to prevent its deletion while we depend on
+it.
+
+To know if your controller needs to implement this test, check the API and see
+if it accepts any `<Resource>Ref` field in its resource spec.
+
+
+The testing pattern goes like this:
+
+1. create the resource with all dependencies satisfied except for one
+    1. check that ORC is waiting on the missing dependency to exist
+1. repeat with another dependency until you've checked all possible dependencies
+1. create the missing dependency
+    1. check that the resource is finally created
+1. delete all dependencies
+    1. validate that ORC prevented the deletion due to a finalizer. There might
+       be exceptions (for example flavor for a server)
+1. delete the resource
+    1. verify that all dependencies are now deleted
+
+#### import
+
+The `import` test validates import filters, and adoption.
+
+The testing pattern goes like this:
+
+1. create an unmanaged resource using all of the available filters for the resource
+    1. verify it is waiting on an external resource matching the filters to exist
+1. create a resource matching all of the import filter except for the name that
+   is a superstring of the one specified in the import filter
+    1. verify that this resource is not being imported -- it validates that we
+       don't perform regex-based name search (some OpenStack projects do it)
+1. create a resource marching all of the import filters, including the name.
+    1. verify that the imported resource is available and the observed status
+       correspond to the one of the created resource
+    1. validate that the previously created resource wasn't imported as the new
+       one, again, because of regex-based name matching
+
+All controllers should implement this test.
+
+#### import-error
+
+In the `import-error` test, we verify that an import filter matching multiple
+resources returns an error.
+
+All controllers should implement this test.
+
+#### import-dependency
+
+The `import-dependency` test is required for resources that allow
+a `<Resource>Ref` field in their import filters.
+
+We want to verify that ORC waits for the dependency to be created and available
+when it creates the resource, but that it allows deleting the dependency
+without the resource being deleted first. This ensures that unmanaged resources
+don't add deletion guards on dependencies.
+
+The pattern for this test is:
```

Comment:
I think it's worth to add an example for this one as it's not trivial (for me)

Reply:
Or link to one in code?

Reply:
I added examples for all test patterns.